File: paimon-core/src/main/java/org/apache/paimon/index/pksorted/PkSortedBucketIndexState.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.index.pksorted;

import org.apache.paimon.index.IndexFileMeta;
import org.apache.paimon.index.pk.PrimaryKeyIndexSourceFile;
import org.apache.paimon.index.pk.PrimaryKeyIndexSourceMeta;
import org.apache.paimon.index.pk.PrimaryKeyIndexSourcePolicy;
import org.apache.paimon.io.DataFileMeta;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.TreeMap;

/** Immutable sorted-index state for one field and bucket. */
public final class PkSortedBucketIndexState {

    private static final class PayloadCandidate {

        private final IndexFileMeta payload;
        private final PkSortedIndexGroup group;
        private final List<PrimaryKeyIndexSourceFile> activeSources;

        private PayloadCandidate(
                IndexFileMeta payload,
                PkSortedIndexGroup group,
                List<PrimaryKeyIndexSourceFile> activeSources) {
            this.payload = payload;
            this.group = group;
            this.activeSources = activeSources;
        }
    }

    private final List<PkSortedIndexGroup> groups;
    private final List<PrimaryKeyIndexSourceFile> coveredSourceFiles;
    private final List<PrimaryKeyIndexSourceFile> uncoveredSourceFiles;
    private final List<IndexFileMeta> rejectedPayloads;

    private PkSortedBucketIndexState(
            List<PkSortedIndexGroup> groups,
            List<PrimaryKeyIndexSourceFile> coveredSourceFiles,
            List<PrimaryKeyIndexSourceFile> uncoveredSourceFiles,
            List<IndexFileMeta> rejectedPayloads) {
        this.groups = Collections.unmodifiableList(groups);
        this.coveredSourceFiles = Collections.unmodifiableList(coveredSourceFiles);
        this.uncoveredSourceFiles = Collections.unmodifiableList(uncoveredSourceFiles);
        this.rejectedPayloads = Collections.unmodifiableList(rejectedPayloads);
    }

    public static PkSortedBucketIndexState fromActiveDataFiles(
            int fieldId,
            String indexType,
            List<DataFileMeta> activeDataFiles,
            List<IndexFileMeta> activePayloads) {
        Map<Integer, List<PrimaryKeyIndexSourceFile>> sourcesByLevel = new TreeMap<>();
        for (DataFileMeta dataFile : activeDataFiles) {
            if (PrimaryKeyIndexSourcePolicy.shouldRead(dataFile)) {
                sourcesByLevel
                        .computeIfAbsent(dataFile.level(), ignored -> new ArrayList<>())
                        .add(
                                new PrimaryKeyIndexSourceFile(
                                        dataFile.fileName(), dataFile.rowCount()));
            }
        }
        for (List<PrimaryKeyIndexSourceFile> sources : sourcesByLevel.values()) {
            sources.sort(Comparator.comparing(PrimaryKeyIndexSourceFile::fileName));
        }

        Map<Integer, List<PayloadCandidate>> candidatesByLevel = new TreeMap<>();
        List<IndexFileMeta> rejected = new ArrayList<>();
        for (IndexFileMeta payload : activePayloads) {
            try {
                PrimaryKeyIndexSourceMeta sourceMeta =
                        PrimaryKeyIndexSourceMeta.fromIndexFile(payload);
                List<PrimaryKeyIndexSourceFile> activeLevelSources =
                        sourcesByLevel.get(sourceMeta.dataLevel());
                if (activeLevelSources == null) {
                    rejected.add(payload);
                    continue;
                }

                List<PrimaryKeyIndexSourceFile> payloadSources = sourceMeta.sourceFiles();
                List<PrimaryKeyIndexSourceFile> activeIntersection =
                        activeIntersection(activeLevelSources, payloadSources);
                if (activeIntersection == null || activeIntersection.isEmpty()) {
                    rejected.add(payload);
                    continue;
                }

                Optional<PkSortedIndexGroup> group =
                        PkSortedIndexGroup.create(
                                fieldId,
                                indexType,
                                payloadSources,
                                Collections.singletonList(payload));
                if (!group.isPresent()) {
                    rejected.add(payload);
                    continue;
                }
                candidatesByLevel
                        .computeIfAbsent(sourceMeta.dataLevel(), ignored -> new ArrayList<>())
                        .add(new PayloadCandidate(payload, group.get(), activeIntersection));
            } catch (RuntimeException ignored) {
                rejected.add(payload);
            }
        }

        List<PkSortedIndexGroup> groups = new ArrayList<>();
        Map<Integer, Set<PrimaryKeyIndexSourceFile>> coveredSourcesByLevel = new TreeMap<>();
        for (Map.Entry<Integer, List<PayloadCandidate>> entry : candidatesByLevel.entrySet()) {
            List<PayloadCandidate> levelCandidates = entry.getValue();
            if (levelCandidates.size() != 1) {
                for (PayloadCandidate candidate : levelCandidates) {
                    rejected.add(candidate.payload);
                }
                continue;
            }
            PayloadCandidate candidate = levelCandidates.get(0);
            groups.add(candidate.group);
            coveredSourcesByLevel
                    .computeIfAbsent(entry.getKey(), ignored -> new HashSet<>())
                    .addAll(candidate.activeSources);
        }

        List<PrimaryKeyIndexSourceFile> covered = new ArrayList<>();
        List<PrimaryKeyIndexSourceFile> uncovered = new ArrayList<>();
        for (Map.Entry<Integer, List<PrimaryKeyIndexSourceFile>> entry :
                sourcesByLevel.entrySet()) {
            Set<PrimaryKeyIndexSourceFile> coveredSources =
                    coveredSourcesByLevel.getOrDefault(entry.getKey(), Collections.emptySet());
            for (PrimaryKeyIndexSourceFile source : entry.getValue()) {
                (coveredSources.contains(source) ? covered : uncovered).add(source);
            }
        }
        return new PkSortedBucketIndexState(groups, covered, uncovered, rejected);
    }

    private static List<PrimaryKeyIndexSourceFile> activeIntersection(
            List<PrimaryKeyIndexSourceFile> activeSources,
            List<PrimaryKeyIndexSourceFile> payloadSources) {
        List<PrimaryKeyIndexSourceFile> intersection = new ArrayList<>();
        int activeSourceIndex = 0;
        for (int i = 0; i < payloadSources.size(); i++) {
            PrimaryKeyIndexSourceFile source = payloadSources.get(i);
            if (i > 0 && payloadSources.get(i - 1).fileName().compareTo(source.fileName()) >= 0) {
                return null;
            }
            while (activeSourceIndex < activeSources.size()
                    && activeSources.get(activeSourceIndex).fileName().compareTo(source.fileName())
                            < 0) {
                activeSourceIndex++;
            }
            if (activeSourceIndex == activeSources.size()
                    || !activeSources.get(activeSourceIndex).fileName().equals(source.fileName())) {
                continue;
            }
            if (activeSources.get(activeSourceIndex).rowCount() != source.rowCount()) {
                return null;
            }
            intersection.add(source);
        }
        return intersection;
    }

    public List<PkSortedIndexGroup> groups() {
        return groups;
    }

    public List<PrimaryKeyIndexSourceFile> coveredSourceFiles() {
        return coveredSourceFiles;
    }

    public List<PrimaryKeyIndexSourceFile> uncoveredSourceFiles() {
        return uncoveredSourceFiles;
    }

    public List<IndexFileMeta> rejectedPayloads() {
        return rejectedPayloads;
    }
}

File: paimon-python/pypaimon/index/pksorted/pk_sorted_bucket_index_state.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from dataclasses import dataclass

from pypaimon.index.pk.primary_key_index_source_file import PrimaryKeyIndexSourceFile
from pypaimon.index.pk.primary_key_index_source_meta import PrimaryKeyIndexSourceMeta
from pypaimon.index.pk.primary_key_index_source_policy import should_read
from pypaimon.index.pksorted.pk_sorted_index_group import PkSortedIndexGroup


@dataclass(frozen=True)
class PkSortedBucketIndexState:
    groups: tuple
    covered_source_files: tuple
    uncovered_source_files: tuple
    rejected_payloads: tuple

    @staticmethod
    def from_active_data_files(field_id, index_type, active_data_files, active_payloads):
        sources_by_level = {}
        for data_file in active_data_files:
            if not should_read(data_file):
                continue
            sources_by_level.setdefault(data_file.level, []).append(
                PrimaryKeyIndexSourceFile(data_file.file_name, data_file.row_count))
        for sources in sources_by_level.values():
            sources.sort(key=lambda source: source.file_name)

        candidates_by_level = {}
        rejected = []
        for payload in active_payloads:
            try:
                source_meta = PrimaryKeyIndexSourceMeta.from_index_file(payload)
                active_level_sources = sources_by_level.get(source_meta.data_level)
                if active_level_sources is None:
                    rejected.append(payload)
                    continue
                active_intersection = _active_intersection(
                    active_level_sources, source_meta.source_files)
                if not active_intersection:
                    rejected.append(payload)
                    continue
                group = PkSortedIndexGroup.create(
                    field_id, index_type, source_meta.source_files, [payload])
                if group is None:
                    rejected.append(payload)
                    continue
                candidates_by_level.setdefault(source_meta.data_level, []).append(
                    (payload, group, active_intersection))
            except (TypeError, ValueError):
                rejected.append(payload)

        groups = []
        covered_sources_by_level = {}
        for level, candidates in sorted(candidates_by_level.items()):
            if len(candidates) != 1:
                rejected.extend(candidate[0] for candidate in candidates)
                continue
            _, group, active_intersection = candidates[0]
            groups.append(group)
            covered_sources_by_level[level] = set(active_intersection)

        covered = []
        uncovered = []
        for level, sources in sorted(sources_by_level.items()):
            covered_sources = covered_sources_by_level.get(level, set())
            for source in sources:
                (covered if source in covered_sources else uncovered).append(source)
        return PkSortedBucketIndexState(
            tuple(groups), tuple(covered), tuple(uncovered), tuple(rejected))


def _active_intersection(active_sources, payload_sources):
    intersection = []
    active_source_index = 0
    for index, source in enumerate(payload_sources):
        if index > 0 and payload_sources[index - 1].file_name >= source.file_name:
            return None
        while (active_source_index < len(active_sources)
               and active_sources[active_source_index].file_name < source.file_name):
            active_source_index += 1
        if (active_source_index == len(active_sources)
                or active_sources[active_source_index].file_name != source.file_name):
            continue
        if active_sources[active_source_index].row_count != source.row_count:
            return None
        intersection.append(source)
    return intersection
